FSE: Update blocks categories

fullofcaffeine opened the pull request:
#### Changes proposed in this Pull Request

- **Update the category names for FSE blocks**. More details: pc6Uok-1b-p2 and [here](https://github.com/Automattic/wp-calypso/issues/43198) (second task listed in the issue). 
-  **Adds a helper function that is used to safely update category names while still supporting old environments**
- **Fix the FSE Jest configuration for TypeScript**. It was needed in order for to properly transpile the test and the helper function included in the changeset when they are consumed by Jest. I decided not to split this change in another PR for two reasons: it's small and it's directly related to the change here.

#### Related PRs
* Jetpack blocks: https://github.com/Automattic/jetpack/pull/16363
* wpcom-blocks: https://github.com/Automattic/wpcom-blocks/pull/178
* block-experiments: https://github.com/Automattic/block-experiments/pull/110

#### Categories

- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-dc263ccf11dec7e5cc2cd31b6672b8dfR35) newspack's `a8c/blog-posts` from `layout` -> `widgets`;
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-05d7ff5647d5a588247b7c0a0bb98bf9R36) newspack's `a8c/posts-carousel` from `layout` -> `widgets`*[0]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-53b5a98f30dc7e6d356d4dae754c02daR24) `a8c/template` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-6893b1bc3b7ff82118bc9334f2e83bbbR26) `a8c/navigation-menu` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-df04a9b006edc7c589a5d70ee6f61e4bR22) `a8c/post-content` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-3b05c15146c5a00d3547626c18b7ffecR19) `a8c/site-title` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-37e1c83f9166b0a7f9aecc29345b4050R22)`a8c/site-credit` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-029c4ccfa1c3a256ee21db826f2cc496R24)  `a8c/site-description` from `layout` -> `design`*[1]
- [X] [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-d8422954d098352028006c46af21bdffR52) `a8c/donations` from `common` -> `widgets`*[0]
- [X] Keep `jetpack/event-countdown` in the `widgets` category
- [X] Keep `jetpack/timeline` in the `widgets` category
- [X] Keep `jetpack/timeline-item` in the `widgets` category
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-e3d4a24841302670cbb36f6358268f27R15) `premium-content/button` from `layout` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-188c53da2d0b0b206b3ffca75cf1029bR15) `premium-content/login-button` from `layout` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-9988e0ed5ce04941b7958abf1bf5f8f5R18) `premium-content/container` from `common` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-f996a37875efc838ad070466ff8eeaf1R18) `premium-content/logged-out-view` from `common` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-15eafa59ca29a72c0c1539f1dc7c5666R14) `premium-content/subscriber-view` from `common` -> `design`

_*[0] This was a guess, let me know if this is the right category to move it to._
_*[1] I did set the category name for those using our helper function, even though I understand that a migration already took care of automatically mapping `layout` to `design` if I'm not mistaken. Let me know if you think it would not be needed for those cases._


#### TODO

- [X] Unit test for the helper function: the test itself is trivial, but I didn't add because I still need to learn more about how unit testing is done within calypso. I'll have a look at it soon.
- [X] Add support for TS for FSE tests.
- [X] Should we write tests to check that each block is assigned to the right category names depending on the set of categories? Not sure what is the testing policy for Calypso and what coverage we aim for. Any insights appreciated! (Jon said it's not needed)
- [x] Squash and rebase from master before merging

#### Outstanding questions

- Do we need to do anything with the category of the [old/deprecated](https://github.com/Automattic/wp-calypso/blob/master/apps/full-site-editing/full-site-editing-plugin/posts-list-block/blocks/posts-list/index.js#L36) `posts-list-block`? (More details in p1593038635493700-slack-cylon) 








#### Testing instructions

##### Atomated tests

* `cd` to `wp-calypso/apps/full-site-editing` and run `yarn test:js`.

##### To manually test the changes

1. Setup a simple FSE-enabled site and atomic ephemeral sites by following the instructions here: PCYsg-ly5-p2. Map it to your sandbox.
2. `cd` to `wp-calypso/apps/full-site-editing` and run `yarn dev --sync` to sync it to your sandbox.
3. Choose a theme that's compatible with FSE (i.e `Exford`)
4. Edit your home-page
5. Add a new block and search for the block you're testing. Verify it's in the right category. 

##### Test the scenarios above in the following environments:**
- [X] Simple website in calypso
- [X] Simple website in wp-admin
- [X] Atomic website
- [X] Atomic website with Gutenberg plugin disabled

~Note: If [this PR](https://github.com/WordPress/gutenberg/pull/23695) gets approved and merged, then I'll remove the helper function and tests from here (including the installation/configuration of jest).~

fullofcaffeine committed:
Change the category of FSE blocks from legacy to the updated ones

Context: https://github.com/Automattic/wp-calypso/issues/43198.

- Also add a helper function for FSE blocks that can be used to check if old category names exist in the store and fall-backs to an older version(s) otherwise. This function should be used when the updated category is a completely new category being introduced by a new version of Gutenberg. You then pass this new category + the older categorie(s) that will be tried in sequence if the previous one was not found.
- Tweaked the FSE Jest configuration to support TypeScript.

a8ci18n commented on apps/full-site-editing/full-site-editing-plugin/newspack-blocks/blog-posts-block-editor.js:
```diff
@@ -31,7 +32,7 @@ addFilter(
 registerBlockType( blockName, {
 	...settings,
 	title: __( 'Blog Posts', 'full-site-editing' ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **80** times:
`translate( 'Blog Posts' )` *ES Score: 10*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/fse-blocks-categories-collections)

sirreal commented on apps/full-site-editing/full-site-editing-plugin/block-helpers.js:
```diff
@@ -0,0 +1,19 @@
+ *
+ * @returns {boolean} value indicating if we're still using old category names.
+ */
+export const hasLegacyCategory = getCategories().some( function ( category ) {
+	return category.slug === 'formatting';
+} );
```
Instead of a capability check you use to build conditionals, we wrap the conditional logic and accept a list of candidate categories?

```ts
function getCategoryWithFallbacks( ...requestedCategories: string[]): string {
	const knownCategories: Array< { slug: string } > = getCategories();
	for ( const requestedCategory of requestedCategories) {
		if ( knownCategories.some( ( { slug } ) => slug === requestedCategory ) ) {
			return requestedCategory;
		}
	}
	// Should we throw an error here? 🤔
	throw new Error('Terrible choice of category 🤦‍♀️');
}

// Usage:
registerBlockType( 'my-block/name', { category: getCategoryWithFallbacks( 'layout', 'widgets' ) } );
```

fullofcaffeine replied:
That's a nice idea! Avoiding having to write a ternary expression would be nice, indeed. Just wondering how we could make this function re-usable across all packages that register blocks. We'll probably need to duplicate it for now?

sirreal replied:
Yep, we'll need to write it once for each package (FSE, Jetpack…).

fullofcaffeine replied:
Btw, I really like this approach as it's basically future proof. If we change categories again, then it's just a matter of passing it as the last one in the argument.

fullofcaffeine replied:
I'm wondering if it makes sense to move this function over to `@wordpress/blocks`... :thinking: this way, we'd avoid the current repetition we have across different block repos.

simison replied:
I like that idea. :-)

WunderBart replied:
> I'm wondering if it makes sense to move this function over to @wordpress/blocks... 🤔 this way, we'd avoid the current repetition we have across different block repos.

IIUC, if we need this to handle new categories on sites where Gutenberg has not been updated yet ([as it handles the other way round automagically](https://github.com/WordPress/gutenberg/pull/23695#issuecomment-653730155)), we can't actually move it to a Gutenberg package, because on those sites that utility won't be available. AFAIR, Jetpack imports use packages that are currently available with the installed version of the Gutenberg plugin, not those that are defined in `package.json` (i.e. `@worpdress/blocks`). Pinging @jeherve here for help 🙏 

If the above is true and we end up dropping the Gutenberg PR, how about we make it into a [Calypso package](https://github.com/Automattic/wp-calypso/tree/master/packages#packages) instead?

fullofcaffeine replied:
> IIUC, if we need this to handle new categories on sites where Gutenberg has not been updated yet (as it handles the other way round automagically), we can't actually move it to a Gutenberg package, because on those sites that utility won't be available

But isn't the `@wordpress/data` a package that can be installed separately (even though it lives within the Gutenberg repo)? Right now, we need this function in the following packages:

- https://github.com/Automattic/wpcom-blocks/pull/178
- https://github.com/Automattic/jetpack/pull/16363
- https://github.com/Automattic/wp-calypso/pull/43670
- https://github.com/Automattic/block-experiments/pull/110

Wouldn't it just be a matter of publishing a new version of `@wordpress/blocks` and then updating it on these packages? 

I'm fine abandoning that PR and moving it to a Calypso package if we find it's the best way, by the way :)

fullofcaffeine replied:
Ah, sorry I read too fast, you actually mention here:

> AFAIR, Jetpack imports use packages that are currently available with the installed version of the Gutenberg plugin, not those that are defined in package.json (i.e. @worpdress/blocks). Pinging @jeherve here for help 

Let's wait for @jeherve's insights about that, then :)

Thanks!

deBhal replied:
Have we considered a linting rule that checks the category in warning about new categories as well? Something like checking the category of the `registerBlockType()` and warning if we know it's a category that's not going to work with older gutenbergs?

sirreal reviewed (commented):
Thanks for starting on this! I know it's a draft, I left some early thoughts.

fullofcaffeine commented on apps/full-site-editing/full-site-editing-plugin/block-helpers.js:
```diff
@@ -0,0 +1,19 @@
+
+/**
+ * Indicates whether or not we're in an environment that is still
+ * using old block [category names]{@link https://github.com/WordPress/gutenberg/pull/19279}.
```
First time I use it actually, it's documented here: https://jsdoc.app/tags-inline-link.html.

fullofcaffeine commented on apps/full-site-editing/full-site-editing-plugin/block-helpers.js:
```diff
@@ -0,0 +1,19 @@
+ * This is not perfect, but works well for checking if we need to register
+ * a block with an old category name or the new one, in order to provide
+ * backwards compatibility with older Gutenberg versions.
+ *
+ * This is the same strategy used by [CoBlocks]{@link https://github.com/godaddy-wordpress/coblocks/pull/1535}.
```
Cool, I'll remove it :+1:

fullofcaffeine commented on apps/full-site-editing/full-site-editing-plugin/block-helpers.js:
```diff
@@ -0,0 +1,19 @@
+/**
```
Awesome, I'll follow your suggestion of abstracting the logic of returning the correct cat name in a function, as you suggested in your comment [below](https://github.com/Automattic/wp-calypso/pull/43670#discussion_r445675309).

sirreal commented on apps/full-site-editing/full-site-editing-plugin/block-helpers/block-helpers.test.ts:
```diff
@@ -0,0 +1,33 @@
+	getCategories: () => [ { slug: 'foobar' }, { slug: 'barfoo' } ],
+} ) );
+
+describe( 'getCategoriyWithFallbacks', () => {
```
Spelling
```suggestion
describe( 'getCategoryWithFallbacks', () => {
```

fullofcaffeine replied:
Oopsie, thanks for catching that :+1:

sirreal commented on apps/full-site-editing/full-site-editing-plugin/block-helpers/block-helpers.ts:
```diff
@@ -0,0 +1,27 @@
+ * @throws {Error} if the no categories could be found.
+ */
+export function getCategoryWithFallbacks( ...requestedCategories: string[] ): string {
+	const knownCategories: Array< { slug: string } > = getCategories();
```
Let's let the type be inferred here, we should have a [good type for `getCategories`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ab9cec4dc372cb96964c6beaf2fa48026c8751dd/types/wordpress__blocks/api/categories.d.ts#L12).
```suggestion
	const knownCategories = getCategories();
```

fullofcaffeine replied:
:+1: Doing some type hygiene was part of my plan btw, but I started with your verbatim copy in order to work on setting up TS for Jest, as it wasn't working before.

sirreal reviewed (commented):
Looking good, I left a few comments.

> This commit has been added but probably will not be included as part of the PR. It's not related to the changeset but was required for me to get my wp-calypso env running.

It looks like there are a few changes to the setup here, it may make sense to break those out into an independent PR. There are multiple concerns here, but the changes are small enough that it's reviewable. At your discretion either way 👍 

> Should we write tests to check that each block is assigned to the right category names depending on the set of categories? Not sure what is the testing policy for Calypso and what coverage we aim for.

Let's stay focused on testing just the added utility function you added. If we start testing the block registration, we quickly get outside of unit tests.

e2e test may make sense at some point to ensure blocks are in the expected categories, but that's not the scope we're looking at here.

fullofcaffeine commented on apps/full-site-editing/bin/babel-transform.js:
```diff
@@ -0,0 +1,8 @@
+/**
```
I don't know if this dir is the best place for this file, although it felt natural to put it at the same level as the file that is consuming it. I'm open to suggestions.

fullofcaffeine replied:
By the way, I'd love some insights about something: FSE uses `calypso-build` which **already** provides some Jest presets. However, it's using presets from `@wordpress/scripts/config/jest-unit.config.js` instead. 

I did try using the presets from Calypso and while the (TS) suite in this changeset worked, I started having errors in other FSE tests, related to `import` being used outside a module context, `React` not being defined and `window` not found (the latest two were easy to fix, for undefined `window`, I just needed to set the `testEnvironment` jest config setting to `jsdom` and for the undefined `React`, I just needed to import `React`). It ended up being more trouble than it's worth, so I just decided to keep the ways things were and just append the `typescript-preset` here for the transformer here. This works well.

sirreal replied:
With this package using `calypso-build` to build, I'd like to use its jest preset: `@automattic/calypso-build`

https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-build/jest-preset.js

fullofcaffeine replied:
That's the same reasoning I used and I tried using this preset, but while TS worked fine, I go additional errors:||

> I did try using the presets from Calypso and while the (TS) suite in this changeset worked, I started having errors in other FSE tests, related to import being used outside a module context, React not being defined and window not found (the latest two were easy to fix, for undefined window, I just needed to set the testEnvironment jest config setting to jsdom and for the undefined React, I just needed to import React).

I'll give it another try.

sirreal replied:
I _very much dislike adding more customization and configuration here_. That's not a critique of this PR or your work, but a critique of the current state of this app's setup.

It has one foot in Calypso build (which has many examples of working _just fine_ with TypeScript in this repo). It has another foot in the `@wordpress/*` side of things (wp-scripts, etc.).

Most of the build seems to be handled by `calypso-build`, but most of the testing seems to be run via `wp-scripts` and its family of helpers. I'd love to see this package commit to one direction or the other, honestly I don't care too much, but committing to either of these tooling solutions would be a better solution.

As I see it right now, we have two alternatives, both of which should work well and remove the amount of configuration we need to include here. However, we've wound up in a situation where we need another complete set of configuration on top of two solutions in order to smooth over the incompatibilities.

😤

I'll set aside my frustrations now because we're not here to solve the setup issues in this PR. I know I started us down this road by suggestions TypeScript, but it's very troubling to me that it doesn't work.

We can ignore these concerns for this PR.

(👋  tooling partner in crime @ockham)

sirreal commented on apps/full-site-editing/full-site-editing-plugin/newspack-blocks/blog-posts-block-editor.js:
```diff
@@ -31,7 +32,7 @@ addFilter(
 	...settings,
 	title: __( 'Blog Posts', 'full-site-editing' ),
-	category: 'layout',
+	category: getCategoryWithFallbacks( 'layout', 'widgets' ),
```
I'd expect our new categories to be listed first. The first registered category will be applied, right?

fullofcaffeine replied:
No, the implementation currently goes sequentially from left to right, checks if the category exists, use it, if not, tries the next one, and so on. What you say makes sense though from a UX perspective, I'll look into changing it.

fullofcaffeine replied:
Yeah, the first registered category will be applied, so you're right, it makes sense to put the new one first. I'll change it.

fullofcaffeine commented:
I decided to [revert](https://github.com/Automattic/wp-calypso/pull/43670/commits/8ba7810ff76e938179a0ec8093c3fd3cfc04ed5a) my changes to the premium block categories because:

1. I wasn't able to test them, they seem to be overridden somewhere. I'm still researching about this but it seems it has to do with https://github.com/Automattic/wp-calypso/tree/master/apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications (see: p1593555242136300-slack-cylon)

2. Even if I figure out how to change them, there's an open issue [here](https://github.com/Automattic/wp-calypso/issues/41579) and the team seems to want to keep it in the "New" category for now and create an "Earn" collection for it. It seems there's not an agreement yet though and from what I see, it's out of the scope of this PR>

simison commented:
Yeah, I'd also just rely on stakeholder team(s) on those blocks and leave them be if they so wish. FYI @lessbloat. :wave:

We can consider this unblocked and proceed with the rest of the blocks then? :-)

fullofcaffeine commented:
Hi Mikael :wave: 

> We can consider this unblocked and proceed with the rest of the blocks then? :-)

I changed all the blocks under FSE I could find (the ones that use `registerBlockType`). Did I miss any? 

The PR for Jetpack is here: https://github.com/Automattic/jetpack/pull/16363

These are the two outstanding group of blocks that need to be dealt with according to the original [issue](https://github.com/Automattic/wp-calypso/issues/43198). Apart from Jetpack and FSE (and Coblocks), what other app/lib/package has blocks that need renaming?

WunderBart reviewed (dismissed):
Found a typo that was making the build fail - fixed it [here](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-d8422954d098352028006c46af21bdffR12).

Started testing it on Simple and AT sites. Simples look good (tested over v8.3.0 that is currently in production) -  the categories seem to be registered as expected. I noticed some issues on my AT site, though:

1. Newspack Blocks (`Blog Posts` & `Post Carousel`) seem to always get the `widgets` category, regardless of the Gutenberg version (pre/post category names change).
2. Some [Premium Content](https://github.com/Automattic/wp-calypso/tree/0797cbc29f1391948b965ea05691717f691b8110/apps/full-site-editing/full-site-editing-plugin/premium-content) blocks are throwing category errors when pre-category-update Gutenberg is used:
   ```
   The block "premium-content/buttons" must have a registered category.
   The block "premium-content/login-button" must have a registered category.
   ```

fullofcaffeine commented:
> Found a typo that was making the build fail - fixed it [here](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-d8422954d098352028006c46af21bdffR12).
> 
> Started testing it on Simple and AT sites. Simples look good (tested over v8.3.0 that is currently in production) - the categories seem to be registered as expected. I noticed some issues on my AT site, though:
> 
>     1. Newspack Blocks (`Blog Posts` & `Post Carousel`) seem to always get the `widgets` category, regardless of the Gutenberg version (pre/post category names change).
> 
>     2. Some [Premium Content](https://github.com/Automattic/wp-calypso/tree/0797cbc29f1391948b965ea05691717f691b8110/apps/full-site-editing/full-site-editing-plugin/premium-content) blocks are throwing category errors when pre-category-update Gutenberg is used:
>        ```
>        The block "premium-content/buttons" must have a registered category.
>        The block "premium-content/login-button" must have a registered category.
>        ```

Thanks a lot for taking the time to review this! :)

>     1. Newspack Blocks (`Blog Posts` & `Post Carousel`) seem to always get the `widgets` category, regardless of the Gutenberg version (pre/post category names change).

As per our Slack convo, for blocks that are updated to default categories that still exist and have existed before, we can do the following:

1. Leave it as is and just let it be registered with the changed category in older GB versions;
2. Reverse the order so that the older category has precedence and is used over the changed version (which will not be used in recent versions because they won't exist).

Any other alternatives? What do you think would be the best way to go here? cc @sirreal 

> Some Premium Content blocks are throwing category errors when pre-category-update Gutenberg is used: 

Hmm, this changeset doesn't include any changes for these blocks. Maybe this is another issue altogether?

WunderBart commented:
> Hmm, this changeset doesn't include any changes for these blocks. Maybe this is another issue altogether?

It looks like those blocks had their categories updated without fallback, i.e. the [Buttons block](https://github.com/Automattic/wp-calypso/blob/master/apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/buttons/index.js#L14).

sirreal commented:
> > 1. Newspack Blocks (`Blog Posts` & `Post Carousel`) seem to always get the `widgets` category, regardless of the Gutenberg version (pre/post category names change).
> >
>
> As per our Slack convo, for blocks that are updated to default categories that still exist and have existed before, we can do the following:
> 
> 1. Leave it as is and just let it be registered with the changed category in older GB versions;

Let's go with this, more stable categories is a _feature_ 🙂

fullofcaffeine commented:
> > Hmm, this changeset doesn't include any changes for these blocks. Maybe this is another issue altogether?
> 
> It looks like those blocks had their categories updated without fallback, i.e. the [Buttons block](https://github.com/Automattic/wp-calypso/blob/master/apps/full-site-editing/full-site-editing-plugin/premium-content/blocks/buttons/index.js#L14).

@bart Thanks for testing this.

I had updated these blocks but decided to revert (commit is no longer available as I squashed it), more details here: https://github.com/Automattic/wp-calypso/pull/43670#issuecomment-652736461.

The category assignment for those two blocks is now using the helper function and passing the legacy `layout` category as the second argument and `design` as the preferred first. I'll list them below for clarity purposes:

* Change `premium-content/button` from `layout` => `design`;
* Change `premium-content/login-button` from `layout` => `design`;

I've left the other two blocks untouched in the legacy `common` category, though:

* Keep `premium-content/container` in `common`;
* Keep `premium-content/subscriber-view` in  `common`.

The `premium-content/container` block seems to have its category overridden to the "New" category, somewhere around [here](https://github.com/Automattic/wp-calypso/tree/master/apps/full-site-editing/full-site-editing-plugin/block-inserter-modifications). I guess this is a temporary thing and that once this override is removed, the one from here will start to apply. However,  `earn` has an [open](https://github.com/Automattic/wp-calypso/issues/41579) issue about that and they are planning to move it over to another category which hasn't been decided yet. @artpi @mmtr Do you guys want to take care of changing the category for `premium-content/container` in a separate PR or would you like us to do it as part of this one? If so, let us know what category we should assign to it, please.

I guess that for `premium-content/subscriber-view`, nothing prevents us from updating its category now, but it doesn't really fit into the new "Text" category (which is the default hard-coded fallback for this category in Gutenberg) and I'm not sure what I would change it to. Maybe `design`? Also, remember its a child block of `premium-content/container`. cc @artpi @mmtr

fullofcaffeine commented:
By the way, while testing the `premium-content/button` and `premium-content/login-button` after changing their categories, I got a "The block has encountered an error and cannot be previewed" error:  

I'm running WordPress 5.4.2 with Twenty Twenty theme, tested with the stock block editor and Gutenberg 8.4.0 (see screenshot below). For FSE, I've tried with this branch and `master` and I got the same error. Am I missing something?

![versions](https://user-images.githubusercontent.com/81248/86981834-4efbf880-c14d-11ea-83c5-da517b6af16d.png)

I guess these blocks shouldn't appear in the picker when the context is the main page since their parent component is the `premium-content/container`? But why is it appearing there then (See screenshot below)?

![wut2](https://user-images.githubusercontent.com/81248/86986902-61c8fa00-c15a-11ea-9f7d-4493e0121798.png)

sirreal commented:
> I've left the other two blocks untouched in the legacy `common` category, though:
> 
> * Keep `premium-content/container` in `common`;
> * Keep `premium-content/subscriber-view` in  `common`.

Nothing should remain in legacy categories (except as fallback). New category will continue to override but we shouldn't rely on it. Let's apply these:

- `premium-content/container` -> `design`
- `premium-content/subscriber-view` -> `widgets`

However, I'm not familiar with these blocks. Pinging @Automattic/earn for suggestions.

> while testing the premium-content/button and premium-content/login-button, I got a The block has encountered an error and cannot be previewed error

I did a quick test and was unable to reproduce the issue. I'll reach out and we'll try to understand what's going on.

sirreal commented on apps/full-site-editing/full-site-editing-plugin/donations/index.js:
```diff
@@ -81,7 +82,7 @@ const settings = {
 		},
 	},
-	category: 'common',
+	category: getCategoryWithFallbacks( 'widgets', 'common' ),
```
Should become just `widgets`, (matching legacy/updated category).

fullofcaffeine replied:
>     1. Newspack Blocks (`Blog Posts` & `Post Carousel`) seem to always get the `widgets` category, regardless of the Gutenberg version (pre/post category names change).
> 

[Done.](https://github.com/Automattic/wp-calypso/pull/43670/commits/f9fec7ae017b908e128d498af11b98ea9b1c95ba)

sirreal commented on apps/full-site-editing/full-site-editing-plugin/newspack-blocks/blog-posts-block-editor.js:
```diff
@@ -31,7 +32,7 @@ addFilter(
 	...settings,
 	title: __( 'Blog Posts', 'full-site-editing' ),
-	category: 'layout',
+	category: getCategoryWithFallbacks( 'widgets', 'layout' ),
```
Same - just `widgets`.

sirreal commented on apps/full-site-editing/full-site-editing-plugin/newspack-blocks/carousel-block-editor.js:
```diff
@@ -28,5 +33,5 @@ addFilter(
 registerBlockType( blockName, {
 	...settings,
-	category: 'layout',
+	category: getCategoryWithFallbacks( 'widgets', 'layout' ),
```
Just `widgets`.

sirreal commented on apps/full-site-editing/tsconfig.json:
```diff
@@ -27,12 +27,16 @@
 		"jsx": "preserve"
 	},
 	"exclude": [ "node_modules" ],
+	// Had to add that in order for vscode to recognize the tests and apply the jest types
+	"include": [
+		"*.test.ts",
+	],
```
Odd… omitting include should include everything. Are you sure this was necessary? Did you do `Restart TS Server`?

This would fail to match most other TypeScript we have in the project as well, which seems like a problem.

fullofcaffeine replied:
Yeah, good point. It could indeed have been because of vscode glitch, I'll check it out again.

fullofcaffeine replied:
If I remove that include, vscode's TS extension starts complaining:

![fooi](https://user-images.githubusercontent.com/81248/87079380-90db7c00-c1eb-11ea-9cdb-c90ad56cd0f5.png)

The interesting thing is that even with including only the test files, TS features are working fine for `block-helper.ts`.

fullofcaffeine replied:
> Looks like an intermittent out-of-memory build error in the desktop job (filed as an issue in #43533):
> 
> ```
> $ check-npm-client && node ${NODE_ARGS:---max_old_space_size=8192} ./node_modules/webpack/bin/webpack.js --config client/webpack.config.js --display errors-only && yarn run build-languages-if-enabled
> events.js:292
>       throw er; // Unhandled 'error' event
>       ^
> Error: write EPIPE
> ```
> 
> Re-starting CI - hopefully that should get it to work.

Copy that :+1:, will keep that in mind for future PRs.

n3f commented:
> > I've left the other two blocks untouched in the legacy `common` category, though:
> > 
> > * Keep `premium-content/container` in `common`;
> > * Keep `premium-content/subscriber-view` in  `common`.
> 
> Nothing should remain in legacy categories (except as fallback). New category will continue to override but we shouldn't rely on it. Let's apply these:
> 
> * `premium-content/container` -> `design`
> * `premium-content/subscriber-view` -> `widgets`
> 
> However, I'm not familiar with these blocks. Pinging @Automattic/earn for suggestions.

When you add the Premium Content Block it populates a container with a subscriber-view and a logged-out-view.  The logged-out-view contains the CTA buttons.  That being said, I don't know how that fits with what you're doing here 🤷 

> 
> > while testing the premium-content/button and premium-content/login-button, I got a The block has encountered an error and cannot be previewed error
> 
> I did a quick test and was unable to reproduce the issue. I'll reach out and we'll try to understand what's going on.

cc @Automattic/serenity

gwwar commented:
Is there additional context for the re-categorization in general? Eg what are the updated guidelines for the Block Editor for types of blocks.

fullofcaffeine commented:
Hi @gwwar :wave:

> Is there additional context for the re-categorization in general? Eg what are the updated guidelines for the Block Editor for types of blocks.

Yes, check our P2 post [here](https://createp2.wordpress.com/2020/07/02/updating-block-categories-cff), and the umbrella issue [here](https://github.com/Automattic/wp-calypso/issues/43198).

Let me know if you have more questions.

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-source".

@ please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/804764). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

gwwar commented:
As an aside, please paraphrase the whys when we can in PR issues. It helps provide context for OSS folks and saves folks a click ;)

nsakaimbo commented:
Looks like an intermittent out-of-memory build error in the desktop job (filed as an issue in #43533):

```
$ check-npm-client && node ${NODE_ARGS:---max_old_space_size=8192} ./node_modules/webpack/bin/webpack.js --config client/webpack.config.js --display errors-only && yarn run build-languages-if-enabled
events.js:292
      throw er; // Unhandled 'error' event
      ^
Error: write EPIPE
 ```

Re-starting CI - hopefully that should get it to work.

fullofcaffeine commented:
> As an aside, please paraphrase the whys when we can in PR issues. It helps provide context for OSS folks and saves folks a click ;)

:+1: Copy that, will keep that in mind for future PRs.

gwwar commented:
🤔 these new categories all seem like odd fits for the earn blocks. I can dig a bit later, but it'd help if we had updated guidance on where to sort blocks from the core project (documentation outside of that PR). For example, are we required to register to a category? Collections look pretty much the same, and I think it'd be fine to just have these appear in the Earn collection.

Some before/after screenshots would help for posterity too in the summary if folks don't mind updating.

fullofcaffeine commented:
Hi @gwwar, thanks a lot for taking the time to look into this!

> thinking these new categories all seem like odd fits for the earn blocks, maybe widgets? I can dig a bit later, but it'd help if we had updated guidance on where to sort blocks from the core project (documentation outside of that PR). For example, are we required to register to a category? Collections look pretty much the same, and I think it'd be fine to just have these appear in the Earn collection.

I might be wrong here, but I think we know as much as you do, hence the call for feedback in the P2 and here.  Regarding your suggestion of using `widgets`, for what blocks specifically? It seems to align with Jon's suggestion here: https://github.com/Automattic/wp-calypso/pull/43670#issuecomment-656154854. Let me know and I'll change accordingly.

> Some before/after screenshots would help for posterity too in the summary if folks don't mind updating.

You mean adding a screenshot for each block in the old category and then in the new category in the category picker?

gwwar commented:
> I might be wrong here, but I think we know as much as you do, hence the call for feedback in the P2 and here. Regarding your suggestion of using widgets, for what blocks specifically?

For the earn related blocks (premium content blocks + donation), they're intended to facilitate purchase flows or display premium content.

> You mean adding a screenshot for each block in the old category and then in the new category in the category picker?

Yes, as a user, what changes would I expect before and after? In this case it'd be some example of how the default inserter looks like with and without the patch applied. Screenshots can help provide context (some people might not know what we're talking about), and aids in documentation on why code changed in the future especially if UI changes a lot (like we see with the gutenberg project).

fullofcaffeine commented:
> Yes, as a user, what changes would I expect before and after? In this case it'd be some example of how the default inserter looks like with and without the patch applied. Screenshots can help provide context (some people might not know what we're talking about), and aids in documentation on why code changed in the future especially if UI changes a lot (like we see with the gutenberg project).

I've added the screenshots here: https://github.com/Automattic/wp-calypso/issues/43198.

gwwar commented:
Chatting with @fullofcaffeine current mappings for Earn blocks are:

premium-content/buttons - design
premium-content/container - common
premium-content/logged-out-view - common
premium-content/login-button - design
premium-content/subscriber-view - common
a8c/donations - widgets

Premium content is pretty specialized but it also involves site layout, so maybe we can map "common" -> "design" here.
I think a8c/donations is a good fit for widget.

I don't really have strong opinions here, so feel free to override my notes if folks can think of a better match.

sirreal commented:
Thanks everyone for helping out!

> premium-content/buttons - design
> premium-content/container - common
> premium-content/logged-out-view - common
> premium-content/login-button - design
> premium-content/subscriber-view - common
> a8c/donations - widgets
> 
> Premium content is pretty specialized but it also involves site layout, so maybe we can map "common" -> "design" here.
> I think a8c/donations is a good fit for widget.

`common` is one of the legacy categories that needs to be replaced with a [new category](https://github.com/WordPress/gutenberg/blob/7442acb6690cd077c153f923d4c4356c9c1ae5ce/packages/blocks/src/store/reducer.js#L34-L41).

For the remaining `common` blocks, I'd propose `design` for all three. These have to do with showing a paywall or paid content, is that correct? That sounds like design more than anything else for me.

- `premium-content/container` - `design`
- `premium-content/logged-out-view` - `design`
- `premium-content/subscriber-view` - `design`

For reference, the valid categories are:
```js
[
	{ slug: 'text', title: __( 'Text' ) },
	{ slug: 'media', title: __( 'Media' ) },
	{ slug: 'design', title: __( 'Design' ) },
	{ slug: 'widgets', title: __( 'Widgets' ) },
	{ slug: 'embed', title: __( 'Embeds' ) },
]
```

fullofcaffeine commented:
> Thanks everyone for helping out!
> 
> > premium-content/buttons - design
> > premium-content/container - common
> > premium-content/logged-out-view - common
> > premium-content/login-button - design
> > premium-content/subscriber-view - common
> > a8c/donations - widgets
> > Premium content is pretty specialized but it also involves site layout, so maybe we can map "common" -> "design" here.
> > I think a8c/donations is a good fit for widget.
> 
> `common` is one of the legacy categories that needs to be replaced with a [new category](https://github.com/WordPress/gutenberg/blob/7442acb6690cd077c153f923d4c4356c9c1ae5ce/packages/blocks/src/store/reducer.js#L34-L41).
> 
> For the remaining `common` blocks, I'd propose `design` for all three. These have to do with showing a paywall or paid content, is that correct? That sounds like design more than anything else for me.
> 
>     * `premium-content/container` - `design`
> 
>     * `premium-content/logged-out-view` - `design`
> 
>     * `premium-content/subscriber-view` - `design`
> 
> 
> For reference, the valid categories are:
> 
> ```js
> [
> 	{ slug: 'text', title: __( 'Text' ) },
> 	{ slug: 'media', title: __( 'Media' ) },
> 	{ slug: 'design', title: __( 'Design' ) },
> 	{ slug: 'widgets', title: __( 'Widgets' ) },
> 	{ slug: 'embed', title: __( 'Embeds' ) },
> ]
> ```

They are already mapped to these categories:

- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-e3d4a24841302670cbb36f6358268f27R15) `premium-content/button` from `layout` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-188c53da2d0b0b206b3ffca75cf1029bR15) `premium-content/login-button` from `layout` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-9988e0ed5ce04941b7958abf1bf5f8f5R18) `premium-content/container` from `common` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-f996a37875efc838ad070466ff8eeaf1R18) `premium-content/logged-out-view` from `common` -> `design`
- [X]  [Change](https://github.com/Automattic/wp-calypso/pull/43670/files#diff-15eafa59ca29a72c0c1539f1dc7c5666R14) `premium-content/subscriber-view` from `common` -> `design`

sirreal commented on apps/full-site-editing/full-site-editing-plugin/block-helpers/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './block-helpers';
```
Let's put the helper function right here, the additional indirection doesn't seem helpful.

fullofcaffeine replied:
:+1:

fullofcaffeine replied:
You're absolutely right, no point in the additional indirection if it's exporting everything from `block-helpers` anyway...

sirreal commented:
Summary of block changes mostly for myself 🙂 

| Block | Previous | New | Observations |
| --- | --- | --- | --- |
| `a8c/donations` | ~`common`~ | `widgets` | Replace legacy category as well (widgets already existed. |
| `a8c/navigation-menu` | `layout` | `design` | |
| `a8c/post-content` | `layout` | `design` | |
| `a8c/site-credit` | `layout` | `design` | |
| `a8c/site-description` | `layout` | `design` | |
| `a8c/site-title` | `layout` | `design` | |
| `a8c/template` | `layout` | `design` | |
| `a8c/blog-posts` | ~`layout`~ | `widgets` | Replace legacy category as well (widgets already existed. |
| `a8c/posts-carousel` | ~`layout`~ | `widgets` | Replace legacy category as well (widgets already existed. |
| `premium-content/buttons` | `design` | `design` | Add fallback `layout` category (this was broken with Gutenberg < 8.4.0?) |
| `premium-content/container` | `common` | `design` |  |
| `premium-content/logged-out-view` | `common` | `design` | |
| `premium-content/login-button` | `design` | `design` | Add fallback `layout` category (this was broken with Gutenberg < 8.4.0?) |
| `premium-content/subscriber-view` | `common` | `design` | |

sirreal commented on apps/full-site-editing/package.json:
```diff
@@ -131,6 +131,9 @@
 		"@types/wordpress__plugins": "^2.3.6",
-		"@wordpress/eslint-plugin": "^6.0.0"
+		"@babel/preset-typescript": "^7.10.1",
+		"@types/jest": "^26.0.3",
```
Let's match root version:

```suggestion
		"@types/jest": "^25.2.1",
```

https://github.com/Automattic/wp-calypso/blob/9a5f2906cc1da31e230efd2000551a176daeb6e0/package.json#L193

sirreal commented on apps/full-site-editing/package.json:
```diff
@@ -131,6 +131,9 @@
+		"@babel/preset-typescript": "^7.10.1",
+		"@types/jest": "^26.0.3",
+		"@wordpress/eslint-plugin": "^6.0.0",
+		"babel-jest": "^26.1.0"
```
Let's match root:
```suggestion
		"babel-jest": "^26.0.1"
```
https://github.com/Automattic/wp-calypso/blob/9a5f2906cc1da31e230efd2000551a176daeb6e0/package.json#L262

sirreal commented on apps/full-site-editing/package.json:
```diff
@@ -131,6 +131,9 @@
 	"devDependencies": {
 		"@types/wordpress__plugins": "^2.3.6",
-		"@wordpress/eslint-plugin": "^6.0.0"
+		"@babel/preset-typescript": "^7.10.1",
```
Let's match calypso-build:

```suggestion
		"@babel/preset-typescript": "^7.8.3",
```

https://github.com/Automattic/wp-calypso/blob/873f41fe772901ee6669fdf48773a4130867d3d9/packages/calypso-build/package.json#L45

sirreal reviewed (commented):
A few small changes, I'll go ahead and apply these…

sirreal commented:
I applied a few changes to align dependency versions across the monorepo. I also rebased.

fullofcaffeine commented on apps/full-site-editing/package.json:
```diff
@@ -133,9 +133,9 @@
+		"@types/jest": "^25.2.1",
 		"@wordpress/eslint-plugin": "^6.0.0",
-		"babel-jest": "^26.1.0"
+		"babel-jest": "^26.0.1"
```
Thanks for the update. Guess it's best to use `lerna add` to add packages as I assume it'd automatically align versions?

sirreal commented on apps/full-site-editing/tsconfig.json:
```diff
@@ -27,12 +27,16 @@
 		// Some type helpers from our webpack/
-		"types": [ "fse" ],
+		"types": [ "fse", "jest" ],
 		"typeRoots": [ "./typings", "./node_modules", "../../node_modules" ],
```
This rule is wrong, it should be 
```suggestion
		"typeRoots": [ "./typings", "./node_modules/@types", "../../node_modules/@types" ],
```

sirreal commented on apps/full-site-editing/tsconfig.json:
```diff
@@ -27,12 +27,16 @@
 		"jsx": "preserve"
 	},
 	"exclude": [ "node_modules" ],
+	// Had to add that in order for vscode to recognize the tests and apply the jest types
+	"include": [
+		"*.test.ts",
+	],
```
With `typeRoots` fixed, this can be removed.
```suggestion
```

fullofcaffeine replied:
Yep, just did this. Thanks!

sirreal commented:
@noahtallen It looks like this is getting a false negative on the [phpcs action](https://github.com/Automattic/wp-calypso/pull/43670/checks?check_run_id=866380670)

It's trying to run phpcs on a file that was added and deleted in this PR, which errors:

```
ERROR: The file "apps/full-site-editing/full-site-editing-plugin/block-helpers/block-helpers.ts" does not exist.
```

sirreal reviewed (approved):
Tested on Atomic. I see the blog posts and post carousel blocks in the updated categories. I think this is good to go 👍 

I'd like to get https://github.com/Automattic/wp-calypso/pull/43670#issuecomment-657745072 sorted out, but I think that's a false negative and unlikely to affect this after merge.

sirreal commented on apps/full-site-editing/tsconfig.json:
```diff
@@ -27,12 +27,12 @@
 		"jsx": "preserve"
 	},
 	"exclude": [ "node_modules" ],
+	// Had to add that in order for vscode to recognize the tests and apply the jest types
```
This comment should be removed.
```suggestion
```

noahtallen commented:
@sirreal I think this makes sense. The issue is that we run phpcs on all "changed" files, which includes deleted files. Will have to adjust that logic to only handle added or changed files

noahtallen commented:
This diff should do it:

```diff
diff --git a/.github/workflows/full-site-editing-plugin.yml b/.github/workflows/full-site-editing-plugin.yml
index c49652d16a..c794789311 100644
--- a/.github/workflows/full-site-editing-plugin.yml
+++ b/.github/workflows/full-site-editing-plugin.yml
@@ -133,11 +133,8 @@ jobs:
       with:
         token: ${{ secrets.GITHUB_TOKEN }}
 
-    - name: Displaying changed files
-      run: echo ${{ join( fromJson( steps.changes.outputs.all ) ) }}
-
     - name: Execute phpcs on changed files
-      run: ./vendor/bin/phpcs --standard=apps/phpcs.xml ${{ join( fromJson( steps.changes.outputs.all ), ' ' ) }}
+      run: ./vendor/bin/phpcs --standard=apps/phpcs.xml ${{ join( fromJson( steps.changes.outputs.modified ), ' ' ) }} ${{ join( fromJson( steps.changes.outputs.added ), ' ' ) }}
       if: ${{ steps.changes.outputs.all != '' }}
     
     - name: No changes found
ddd

fullofcaffeine committed:
Fix workflow logic to not run phpcs for deleted files

fullofcaffeine commented:
> Tested on Atomic. I see the blog posts and post carousel blocks in the updated categories. I think this is good to go +1
> 
> I'd like to get [#43670 (comment)](https://github.com/Automattic/wp-calypso/pull/43670#issuecomment-657745072) sorted out, but I think that's a false negative and unlikely to affect this after merge.

@Automattic/good-mountain Fellows, I ended up getting into a :rabbit2: :hole: and couldn't deploy today. But it was for a good cause (hopefully!). It's related to the previous report I made [here](https://github.com/Automattic/wp-calypso/pull/43670#issuecomment-655843442). More details below.

It seems some earn (premium) blocks were broken for a while, and we didn't know:

1. In `master`'s FSE, when using older versions of Gutenberg, the `premium-content/login-button` and `premium-content/buttons` were not appearing in the picker's `design` category is not available there. The `premium-content/content` can be found and added, but one of its sub-blocks appears erro'ed.
1. In `master`'s FSE, when using a recent (> `8.30`) version of Gutenberg, the blocks are shown correctly in the picker but 1)  `premium-content/buttons` fails upon being added to the page 2)  `premium-content/login-button` works 3) `premium-content/content` can be found and inserted, but one of its sub-blocks renders an error message.

The difference when testing with this branch's FSE is that in older Gutenberg versions where the `design` category is not available, the blocks are registered correctly in the legacy category provided so that they are found in the picker. But they err when added to the page. When using more recent versions of Gutenberg, the same thing I described in point 2 above happens.

It seems these issues are independent of the changes added here. Still, the changes here make the two "problematic" premium blocks appear on Gutenberg versions < `8.3.0`. After discussing this with @gwwar and @deBhal, I decided to postpone the deploy for now. More details [here](https://a8c.slack.com/archives/C013QHLF28Y/p1594677291077000).

@deBhal is gardening, and this bug probably falls on gardeners' plate. We spent some time pairing on this today, and he's still looking into it as I end my day now. :australia: 

I could have managed to deploy this today, but I wanted to wait to understand what it's about and if we need to do anything in the category side. I'm pretty sure that this can be merged and perhaps deployed tomorrow.

_*with the help of @deBhal, who kindly offered to help me deploy this PR and paired with me on this issue, thanks a lot!

sirreal commented:
Conversation on handling broken blocks: pbAok1-1ez-p2

gwwar reviewed (approved):
Thanks for shepherding the update @fullofcaffeine! This tests well for Premium Content blocks on Simple Sites (edge and 8.3). 

@Automattic/serenity is looking into potential breakages into Atomic from Premium Content, but it shouldn't affect this PR. We'll follow up shortly.

<img width="1257" alt="Screen Shot 2020-07-14 at 3 08 38 PM" src="https://user-images.githubusercontent.com/1270189/87481314-ed6fda00-c5e3-11ea-8ccb-c857323fbe78.png">
<img width="1257" alt="Screen Shot 2020-07-14 at 3 08 43 PM" src="https://user-images.githubusercontent.com/1270189/87481319-eea10700-c5e3-11ea-8fa2-d3556b6f59c6.png">

fullofcaffeine committed:
Remove fallback to layout from button and login-button from premium-content

The fallback was causing these blocks to appear in older versions of Gutenberg where the `design` category was not available. Until here, all good. However, they were all er'ing upon insertion. We thought it would be better to rollback this change for now.
